File: src/RNetPi.Core/Packets/PacketS2CSourceProperty.cs
using RNetPi.Core.Constants;

namespace RNetPi.Core.Packets;

/// <summary>
/// Server -> Client
/// ID = 0x34
/// Source Property
/// Data:
///     (Unsigned Char) Source ID
///     (Unsigned Char) Property ID
///     (Variable) Property Value
/// </summary>
public class PacketS2CSourceProperty : PacketS2C
{
    public const byte ID = 0x34;

    public PacketS2CSourceProperty(byte sourceID, byte propertyID, object? propertyValue)
    {
        Writer.Write(sourceID);
        Writer.Write(propertyID);
        
        switch (propertyID)
        {
            case SourceProperties.AutoOff:
            case SourceProperties.OverrideName:
                Writer.Write((byte)((propertyValue is bool boolValue && boolValue) ? 0x01 : 0x00));
                break;
            case SourceProperties.AutoOnZones:
                if (propertyValue is IEnumerable<(byte, byte)> zones)
                {
                    foreach (var (controllerID, zoneID) in zones)
                    {
                        Writer.Write(controllerID);
                        Writer.Write(zoneID);
                    }
                }
                break;
        }
    }

    public override byte GetID() => ID;
}

File: src_js/server/packets/PacketS2CSourceProperty.js
const PacketS2C = require("./PacketS2C");
const SourceProperty = require("../../rnet/SourceProperty");

/**
 * Server -> Client
 * ID = 0x34
 * SourceProperty
 * Data:
 *      (Unsigned Char) Update Version
 */
class PacketS2CSourceProperty extends PacketS2C {
    constructor(sourceID, propertyID, propertyValue) {
        super();
        this._buffer.writeUInt8(sourceID);
        this._buffer.writeUInt8(propertyID);
        switch (propertyID) {
            case SourceProperty.PROPERTY_AUTO_OFF:
            case SourceProperty.PROPERTY_OVERRIDE_NAME:
                this._buffer.writeUInt8(propertyValue === true ? 0x01 : 0x00);
                break;
            case SourceProperty.PROPERTY_AUTO_ON_ZONES:
                for (let zoneID of propertyValue) {
                    this._buffer.writeUInt8(zoneID[0]);
                    this._buffer.writeUInt8(zoneID[1]);
                }
                break;
        }
    }

    getID() {
        return 0x34;
    }
}

module.exports = PacketS2CSourceProperty;
